lib/phoenix/pubsub/dns_cluster.ex

```diff
@@ -0,0 +1,139 @@
+defmodule Phoenix.PubSub.DNSCluster do
+  @moduledoc """
+  Simple DNS based cluster discovery.
+
+  A DNS query is made every `:interval` milliseconds to discover new ips.
+  Nodes will only be joined if their node basename matches the basename of the
+  current node. For example if `node()` is `myapp-123@fdaa:1:36c9:a7b:198:c4b1:73c6:1`,
+  a `Node.connect/1` attempt will be made against every IP returned by the DNS query,
+  but will only be successful if there is a node running on the remote host with the same
+  basename, for example `myapp-123@fdaa:1:36c9:a7b:198:c4b1:73c6:2`. Nodes running on
+  remote hosts, but with different basenames will fail to connect and will be ignored.
+
+  ## Examples
+
+  To start in your supervision tree, add the child:
+
+      children = [
+        ...,
+        {Phoenix.PubSub.DNSCluster, name: MyApp.Cluster, query: "myapp.internal"}
+      ]
+
+  See the `start_link/1` docs for all available options.
+
+  If you require more advanced clustering options and strategies, see the
+  [libcluster](https://hexdocs.pm/libcluster) library.
+  """
+  use GenServer
+  require Logger
+
+  defmodule Resolver do
```

Comment:
Why is it a separate module? 🤔

Reply:
just for testing. Alternative is doc false'ing everything on the public DNSCliter